bark_ml/observers/graph_observer.py

```diff
@@ -0,0 +1,554 @@
+    adj_start_idx = n_nodes * n_features
+    adj_end_idx = adj_start_idx + n_nodes ** 2
+    A = tf.reshape(obs[:, adj_start_idx:adj_end_idx], [batch_size, n_nodes, n_nodes])
+
+    if dense:
+      # in dense mode, the nodes of all graphs are 
+      # concatenated into one list that contains all 
+      # feature vectors of the whole batch.
+      # the assignment to graphs is handled by the
+      # node_to_graph_map.
+      F = tf.reshape(F, [batch_size * n_nodes, n_features])
+
+      # find non-zero elements in the adjacency matrix (edges)
+      # and collect their indices
+      A = tf.where(tf.greater(A, 0))
+
+      # we need the indices of the source and target nodes to
+      # be represented as their indices in the whole batch,
+      # in other words: each node index must be the index
+      # of the graph in the batch plus the index of the node
+      # in the graph. E.g. if each graph has 5 nodes, the 
+      # node indices are: graph 0: 0-4, graph 1: 5-9, etc.
+
+      # compute a tensor where each element
+      # is the graph index of the node that is represented
+      # by the same index in A
+      graph_indices = tf.reshape(A[:, 0], [1, -1])
+      graph_indices = tf.scalar_mul(n_nodes, graph_indices)
+      graph_indices = tf.transpose(tf.tile(graph_indices, [2, 1]))
+      
+      A = A[:, 1:] + graph_indices
+      A = tf.cast(A, tf.int32)
+
+      # since the node feature vectors are all stored in a 2D array,
+      # we need to map them to the graphs that they belong to.
+
+      # construct a mapping from the node feature vectors in F
+      # to the graphs in the batch, i.e a list where each element 
+      # contains the index of the graph that the feature vector at
+      # the same index in F belongs to, e.g. if each graph contains
+      # two nodes, then the node_to_graph_map is [0, 0, 1, 1, 2, 2, ...]
+      node_to_graph_map = tf.reshape(tf.range(batch_size), [-1, 1])
+      node_to_graph_map = tf.tile(node_to_graph_map, [1, n_nodes])
+      node_to_graph_map = tf.reshape(node_to_graph_map, [-1])
+
+      return F, A, node_to_graph_map
+
+    # extract edge features
+    n_edge_features = graph_dims[2]
+    E_shape = [batch_size, n_nodes, n_nodes, n_edge_features]
+    E = tf.reshape(obs[:, adj_end_idx:], E_shape)
+
+    return F, A, E
+
+  def _preprocess_agents(self, world):
+    """
+    Preproccesses the agents for constructing an
+    observation by resorting and pruning the 
+    original list of agents.
+
+    Args:
+      world: A `bark.core.World` instance.
+
+    Returns:
+      A list of tuples, consisting of an index and an 
+      agent object element.
+
+      The first element always represents the ego agent.
+      The remaining elements resemble other agents, up 
+      to the limit defined by `self._agent_limit`, sorted 
+      in ascending order with respect to the agents'
+      distance in the world to the ego agent.
+    """
+    ego_agent = world.ego_agent
```

Comment:
we only have the ego_agent available in the observed world

Reply:
I'm not sure I understand, could you please elaborate?

Reply:
This was referring to the comment above. I have seen this before. all good!